source/common/common/packed_struct.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include "source/common/common/assert.h"
+#include "source/common/common/utility.h"
+#include "source/common/protobuf/utility.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+/**
+ * These are helper macros to pack elements of the same type into a struct like
+ * object, with the expectation that some or most of the elements are not set.
```

Comment:
I was thinking about this kind of issue for the router config, per discussion with @yurykats .

However what I was thinking was that what we really want is a map, so that if I have 10 optional strings which are mostly never specified, I can make either a `flat_hash_map<id, string>`, or maybe a unique_ptr<flat_hash_map<id, string>>`. That way the cost for all the unspecified optional things is 8 bytes, and the incremental cost for populating them is pretty small also.

It looks to me like you are saving a unique_ptr for every element here and we can do better when none/few of them are populated.

Reply:
Also we should use InlineStringPtr if we want to reduce storage cost when they *are* specified, and and don't need to mutate the strings.

Reply:
@yurykats FYI

Reply:
`[flat_hash_map](https://abseil.io/docs/cpp/guides/container#:~:text=The%20container%20uses,for%20our%20estimates.)` is probably not efficient enough.

Reply:
in terms of memory footprint? I can believe that.

std::map might actually better (or the absl btree_map version)

Reply:
also `std::unique_ptr<map<>>` should be considered. But if we are really talking about a handful of elements max, then a linked list might be the best approach

struct Entry {
  Entry* next{nullptr};
  KeyEnum name;
  InlineStringPtr value;
};

That will cost 8 bytes for the Entry* in the config structure, and an extra 16 bytes overhead per populated item.

Reply:
if you do the above we'd want the API you present to return value as an optional of some flavor, rather than having a separate has_xxx() and xxx() interface which would each cost a list-walk.

Reply:
With `PACKED_DATA` as opposed to linked list, you don't pay the cost of `next` for every element (only once for the first). Also, you avoid the padding after `name` as all indexes are at the end and one byte each. Lastly, you get constant access times. We could have `T` be `InlineStringPtr` instead of `std::string` for `PACKED_DATA`.

What are your thoughts about that?

Having said that, I'm not a fan of macros but don't see a better way of reducing memory usage further.

Reply:
What's the footprint for PACKED_DATA when nothing is populated? That's the most common case, right?

Reply:
It's 8 + n + 2 bytes + padding, where n is the maximum number of elements (If my calculation is correct). So for n = 5 for example, it's 16 bytes.

Reply:
well the linked list only has 8 bytes regardless of n :)

I don't understand how you get 8+n+2bytes+padding. I see this in the PR:
```
std::unique_ptr<T[]> data(capacity > 0 ? new T[capacity] : nullptr);
```

Doesn't that mean T*n ?

Reply:
The way to use it is to first calculate the number of non-empty elements, and initialise it with that number. If all elements are empty, `capacity` will be 0.

Reply:
Per IM I am really warming up to this data layout and have asked Pradeep to see if he can build a usable API w/o macros :)

/wait